### PR TITLE
Safety checks in SYCL abstraction layer

### PR DESCRIPTION
**Pull request by rlnx**
- Meaningful changes is in `cpp/daal/include/services/internal/sycl` and `cpp/daal/include/services/internal/buffer*.h`
- Other modifications are minor and due to changes in the abstraction layer interface
- Added "more safe" options to cpp_sycl examples makefile, align options in other makefiles
- Fix warnings caused by new options in examples

### CODE REVIEW

File: cpp/daal/include/services/internal/sycl/math/blas_executor.h [hunk 1]
```diff
@@ -200,10 +208,14 @@ class SyrkExecutor
...
     static void run(cl::sycl::queue & queue, const math::UpLo upper_lower, const math::Transpose trans, const size_t n, const size_t k,
                     const double alpha, const UniversalBuffer & a_buffer, const size_t lda, const size_t offsetA, const double beta,
-                    UniversalBuffer & c_buffer, const size_t ldc, const size_t offsetC, services::Status & status)
+                    UniversalBuffer & c_buffer, const size_t ldc, const size_t offsetC, Status & status)
     {
-        Execute op(queue, upper_lower, trans, n, k, alpha, a_buffer, lda, offsetA, beta, c_buffer, ldc, offsetC, status);
-        TypeDispatcher::floatDispatch(a_buffer.type(), op);
+        DAAL_ASSERT(!a_buffer.empty());
+        DAAL_ASSERT(!c_buffer.empty());
+        DAAL_ASSERT(a_buffer.type() == c_buffer.type());
```
Comment (amgrigoriev) [inline cpp/daal/include/services/internal/sycl/math/blas_executor.h:215]: buffer size check?

File: cpp/daal/include/services/internal/sycl/math/blas_executor.h [hunk 2]
```diff
@@ -260,10 +271,14 @@ class AxpyExecutor
...
 public:
     static void run(cl::sycl::queue & queue, const uint32_t n, const double a, const UniversalBuffer x_buffer, const int incx,
-                    UniversalBuffer y_buffer, const int incy, services::Status & status)
+                    UniversalBuffer y_buffer, const int incy, Status & status)
     {
-        Execute op(queue, n, a, x_buffer, incx, y_buffer, incy, status);
-        TypeDispatcher::floatDispatch(x_buffer.type(), op);
+        DAAL_ASSERT(!x_buffer.empty());
+        DAAL_ASSERT(!y_buffer.empty());
+        DAAL_ASSERT(x_buffer.type() == y_buffer.type());
```
Comment (amgrigoriev) [inline cpp/daal/include/services/internal/sycl/math/blas_executor.h:278]: Buffer size check?

File: cpp/daal/include/services/internal/sycl/math/lapack_executor.h [hunk 3]
```diff
@@ -65,16 +64,15 @@ class PotrfExecutor
...
+            : queue(queue), uplo(uplo), n(n), a_buffer(a_buffer), lda(lda)
         {}
 
         template <typename T>
-        void operator()(Typelist<T>)
+        void operator()(Typelist<T>, Status & status)
         {
+            DAAL_ASSERT_UNIVERSAL_BUFFER_TYPE(a_buffer, T);
+
             auto a_buffer_t = a_buffer.template get<T>();
```
Comment (amgrigoriev) [inline cpp/daal/include/services/internal/sycl/math/lapack_executor.h:77]: Size check?

File: cpp/daal/include/services/internal/sycl/math/lapack_executor.h [hunk 4]
```diff
@@ -112,16 +111,18 @@ class PotrsExecutor
...
 
         template <typename T>
-        void operator()(Typelist<T>)
+        void operator()(Typelist<T>, Status & status)
         {
+            DAAL_ASSERT_UNIVERSAL_BUFFER_TYPE(a_buffer, T);
+            DAAL_ASSERT_UNIVERSAL_BUFFER_TYPE(b_buffer, T);
+
             auto a_buffer_t = a_buffer.template get<T>();
             auto b_buffer_t = b_buffer.template get<T>();
```
Comment (amgrigoriev) [inline cpp/daal/include/services/internal/sycl/math/lapack_executor.h:128]: size check?

File: cpp/daal/include/services/internal/sycl/math/lapack_executor.h [hunk 5]
```diff
@@ -137,10 +138,14 @@ class PotrsExecutor
...
     static void run(cl::sycl::queue & queue, const math::UpLo uplo, const size_t n, const size_t ny, UniversalBuffer & a_buffer, const size_t lda,
-                    UniversalBuffer & b_buffer, const size_t ldb, services::Status & status)
+                    UniversalBuffer & b_buffer, const size_t ldb, Status & status)
     {
-        Execute op(queue, uplo, n, ny, a_buffer, lda, b_buffer, ldb, status);
-        TypeDispatcher::floatDispatch(a_buffer.type(), op);
+        DAAL_ASSERT(!a_buffer.empty());
+        DAAL_ASSERT(!b_buffer.empty());
+        DAAL_ASSERT(a_buffer.type() == b_buffer.type());
+
```
Comment (amgrigoriev) [inline cpp/daal/include/services/internal/sycl/math/lapack_executor.h:146]: Size check?

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/risrafil-abstraction-layer-sdl-2

File: cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h [hunk 6]
```diff
@@ -76,7 +76,13 @@ class DAAL_EXPORT SyclHomogenNumericTable : public SyclNumericTable
                                                                           services::Status * stat = NULL)
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/daal/include/data_management/data/internal/numeric_table_sycl_homogen.h:76]: @michael-smirnov is addressing issues in data management

File: cpp/daal/include/services/internal/sycl/math/lapack_executor.h (hunk 5 above)
Comment (rlnx) [reply cpp/daal/include/services/internal/sycl/math/lapack_executor.h:146]: Size check is going to be performed in Executor

File: cpp/daal/include/services/internal/sycl/math/blas_executor.h (hunk 1 above)
Comment (rlnx) [reply cpp/daal/include/services/internal/sycl/math/blas_executor.h:236]: This check is going to be performed in `Executor`

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/risrafil-abstraction-layer-sdl-2

**Review by amgrigoriev — APPROVED** [review]

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/risrafil-abstraction-layer-sdl-2

**Comment by rlnx** [comment]
/intelci: run
set infra_branch=dev/risrafil-abstraction-layer-sdl-2

**Pull request merged**